packages/loopring_v3/circuit/Circuits/AmmUpdateCircuit.h

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+#ifndef _AMMUPDATECIRCUIT_H_
+#define _AMMUPDATECIRCUIT_H_
+
+#include "Circuit.h"
+#include "../Utils/Constants.h"
+#include "../Utils/Data.h"
+
+#include "ethsnarks.hpp"
+#include "utils.hpp"
+
+using namespace ethsnarks;
+
+namespace Loopring
+{
+
+// Updates the AMM fee bips and token weights for an account.
+// Also makes a token balance of an account available in the data-availability data.
+class AmmUpdateCircuit : public BaseTransactionCircuit
+{
+  public:
+    // Inputs
+    DualVariableGadget owner;
+    DualVariableGadget accountID;
+    DualVariableGadget tokenID;
+    DualVariableGadget feeBips;
+    DualVariableGadget tokenWeight;
+    DualVariableGadget nonce;
+    DualVariableGadget balance;
+
+    // Increase the nonce
+    AddGadget nonce_after;
+    // Increase the number of conditional transactions
+    UnsafeAddGadget numConditionalTransactionsAfter;
+
+    AmmUpdateCircuit( //
+      ProtoboardT &pb,
+      const TransactionState &state,
+      const std::string &prefix)
+        : BaseTransactionCircuit(pb, state, prefix),
+
+          // Inputs
+          owner(pb, state.accountA.account.owner, NUM_BITS_ADDRESS, FMT(prefix, ".owner")),
+          accountID(pb, NUM_BITS_ACCOUNT, FMT(prefix, ".accountID")),
+          tokenID(pb, NUM_BITS_TOKEN, FMT(prefix, ".tokenID")),
+          feeBips(pb, NUM_BITS_AMM_BIPS, FMT(prefix, ".feeBips")),
+          tokenWeight(pb, NUM_BITS_AMOUNT, FMT(prefix, ".tokenWeight")),
+          nonce(pb, state.accountA.account.nonce, NUM_BITS_NONCE, FMT(prefix, ".nonce")),
+          balance(pb, state.accountA.balanceS.balance, NUM_BITS_AMOUNT, FMT(prefix, ".balance")),
+
+          // Increase the nonce
+          nonce_after(
```

Comment:
It' can currently only be approved onchain, no offchain signature is possible.

First I did it similarly to deposits, but I changed it because we have to make sure the request is unique somehow (which the nonce easily does) and we need to be able to have multiple requests open at the same time to not run into strange issues. Unlike deposits, we can't just process the AMM updates partially or only the latest data, so we need to store data for each request separately, otherwise it's impossible to have multiple requests open/account which can give all sorts of problems when the pool needs to be updated multiple times. That's why I think it's much easier to just do the same as we do for other requests and not do anything we don't do for any other request types, withdrawals included, which increase the nonce as well, so it's actually more uniform with other transactions. The good thing about the current system is that the contract doesn't even have to worry about nonces at all, these come directly from the block data so no need to track them onchain.

> and we implicit restrict the owner of any Amm account to be a contract address.

Why does this matter for the protocol? An Ethereum address is an Ethereum address? The requests don't even have to be processed so it's up to the relayer what it wants to do.